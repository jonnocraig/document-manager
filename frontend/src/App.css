.pageTitle {
  margin-top:20px;
}
