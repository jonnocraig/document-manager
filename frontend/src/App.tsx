import * as React from 'react';
import { Container, Row, Col } from 'reactstrap';
import { IDocument } from './types';
import Nav from './components/navbar';
import List from './components/list';
import Search from './components/search';
import UploadButton from './components/uploadButton';
import './App.css';

export interface IAppProps {
  documents:IDocument[] | null,
  filter:string,
  loading:boolean
}

export interface IAppDispatchProps {
  handleGetDocs:() =>void,
  handleFilterDocs:(filter:string) => void,
  handleDeleteDoc:(id:number) => void,
  handleUploadDoc:(data:FormData) =>void
}

export interface IAppState {
  filter:string
}

export class App extends React.PureComponent<IAppProps & IAppDispatchProps, IAppState> {
  
  constructor(props: IAppProps & IAppDispatchProps) {
    super(props);
    this.state = {
      filter: this.props.filter
    };
  }

  componentDidMount() {
    this.props.handleGetDocs();
  }

  onFilterDocs = (evt:any) => {
    this.props.handleFilterDocs(evt.target.value);
  }

  onUploadDocument = (evt:any) => {
    evt.preventDefault()
    const { files } = evt.target

    if (!files.length) { return; };

    const data = new FormData();
    data.append('filename', files[0]);
    this.props.handleUploadDoc(data);
  }

  render() {
    const { loading, filter, documents, handleDeleteDoc } = this.props;

    return (
      <React.Fragment>
        <Nav />
        <Container>
          <Row>
            <Col sm={{ size: 10, offset: 1}}>
              <h2 className="pageTitle">Document manager</h2>
            </Col>
          </Row>
          <Row>
            <Col sm={{ size: 4, offset: 1}}>
              <Search filter={filter} handleFilter={this.onFilterDocs} />
            </Col>
            <Col sm={{ size: 4, offset: 1}}>
              <UploadButton handleUpload={this.onUploadDocument} />
            </Col>
          </Row>          
          <Row>
            <Col sm={{ size: 10, offset: 1}}>
              { loading ? 
                <p id="loading">Loading...</p> :          
                <List filter={filter} documents={documents} handleDelete={handleDeleteDoc} />
              }
            </Col>
          </Row>
        </Container>
      </React.Fragment>      
    );
  }
}

export default App;
