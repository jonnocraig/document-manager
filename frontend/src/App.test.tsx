import * as React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import List from './components/list';

function setup(nextProps = {}) {
  const props = {
    documents:null,
    filter:'',
    loading:false,
    handleGetDocs: jest.fn(),
    handleFilterDocs:jest.fn(),
    handleDeleteDoc:jest.fn(),
    handleUploadDoc:jest.fn(),
    ...nextProps
  };
  const wrapper = shallow(<App {...props} />);

  return {
    props,
    wrapper
  };
}

const DOCUMENT = {
  id: 1,
  filename:'123test',
  originalFilename:'test',
  fileSize:1.23,
  fileExtension:'.txt',
  created_at:'01 Jan 2018',
  updated_at:'01 Jan 2018'
};

describe('<App>', () => {
  it('should match the snapshot', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleGetDocs() in componentDidMount', () => {
    const { props } = setup();
    expect(props.handleGetDocs.mock.calls.length).toBe(1);
  });

  it('should render the document list after fetching data', () => {
    const { wrapper } = setup({ loading: false, data: [DOCUMENT] })
    expect(wrapper.find(List)).toHaveLength(1);
  });

  it('should render loading text if the loading flag is true', () => {
    const { wrapper } = setup({ loading: true, data: [DOCUMENT] })
    expect(wrapper.find('#loading')).toHaveLength(1);
  });

  it('should fire handleFilterDocs with the correct filter when the onFilterDocs method is called', () => {
    const { props, wrapper } = setup();
    const wrapperInstance = wrapper.instance() as App;
    const FILTER = 'abc';
    const event = { preventDefault: jest.fn(), target: { value: FILTER } };
    
    wrapperInstance.onFilterDocs(event);
    wrapper.update();
    expect(props.handleFilterDocs.mock.calls.length).toBe(1);
    expect(props.handleFilterDocs.mock.calls[0][0]).toEqual(FILTER);
  });

  it('should fire handleUploadDoc with the correct Document the onUploadDocument method is called', () => {
    const { props, wrapper } = setup();
    const wrapperInstance = wrapper.instance() as App;
    
    const data = new FormData();
    const DOC = new Blob();
    data.append('filename', DOC);
    const event = { preventDefault: jest.fn(), target: { files: [DOC] } };
    
    wrapperInstance.onUploadDocument(event);
    wrapper.update();
    expect(props.handleUploadDoc.mock.calls.length).toBe(1);
    expect(JSON.stringify(props.handleUploadDoc.mock.calls[0][0])).toMatch(JSON.stringify(data));
  });
});