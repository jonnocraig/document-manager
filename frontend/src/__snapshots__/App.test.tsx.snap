// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    documents={null}
    filter=""
    handleDeleteDoc={[MockFunction]}
    handleFilterDocs={[MockFunction]}
    handleGetDocs={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": undefined,
      }
    }
    handleUploadDoc={[MockFunction]}
    loading={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Nav />,
        <Container
          tag="div"
        >
          <Row
            tag="div"
          >
            <Col
              sm={
                Object {
                  "offset": 1,
                  "size": 10,
                }
              }
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <h2
                className="pageTitle"
              >
                Document manager
              </h2>
            </Col>
          </Row>
          <Row
            tag="div"
          >
            <Col
              sm={
                Object {
                  "offset": 1,
                  "size": 4,
                }
              }
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <SearchButton
                filter=""
                handleFilter={[Function]}
              />
            </Col>
            <Col
              sm={
                Object {
                  "offset": 1,
                  "size": 4,
                }
              }
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <UploadButton
                handleUpload={[Function]}
              />
            </Col>
          </Row>
          <Row
            tag="div"
          >
            <Col
              sm={
                Object {
                  "offset": 1,
                  "size": 10,
                }
              }
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <List
                documents={null}
                filter=""
                handleDelete={[MockFunction]}
              />
            </Col>
          </Row>
        </Container>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Row
              tag="div"
            >
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 10,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <h2
                  className="pageTitle"
                >
                  Document manager
                </h2>
              </Col>
            </Row>,
            <Row
              tag="div"
            >
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 4,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <SearchButton
                  filter=""
                  handleFilter={[Function]}
                />
              </Col>
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 4,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <UploadButton
                  handleUpload={[Function]}
                />
              </Col>
            </Row>,
            <Row
              tag="div"
            >
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 10,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <List
                  documents={null}
                  filter=""
                  handleDelete={[MockFunction]}
                />
              </Col>
            </Row>,
          ],
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 10,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <h2
                  className="pageTitle"
                >
                  Document manager
                </h2>
              </Col>,
              "tag": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <h2
                  className="pageTitle"
                >
                  Document manager
                </h2>,
                "sm": Object {
                  "offset": 1,
                  "size": 10,
                },
                "tag": "div",
                "widths": Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Document manager",
                  "className": "pageTitle",
                },
                "ref": null,
                "rendered": "Document manager",
                "type": "h2",
              },
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 4,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <SearchButton
                    filter=""
                    handleFilter={[Function]}
                  />
                </Col>,
                <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 4,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <UploadButton
                    handleUpload={[Function]}
                  />
                </Col>,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <SearchButton
                    filter=""
                    handleFilter={[Function]}
                  />,
                  "sm": Object {
                    "offset": 1,
                    "size": 4,
                  },
                  "tag": "div",
                  "widths": Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "filter": "",
                    "handleFilter": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <UploadButton
                    handleUpload={[Function]}
                  />,
                  "sm": Object {
                    "offset": 1,
                    "size": 4,
                  },
                  "tag": "div",
                  "widths": Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "handleUpload": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 10,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <List
                  documents={null}
                  filter=""
                  handleDelete={[MockFunction]}
                />
              </Col>,
              "tag": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <List
                  documents={null}
                  filter=""
                  handleDelete={[MockFunction]}
                />,
                "sm": Object {
                  "offset": 1,
                  "size": 10,
                },
                "tag": "div",
                "widths": Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "documents": null,
                  "filter": "",
                  "handleDelete": [MockFunction],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Nav />,
          <Container
            tag="div"
          >
            <Row
              tag="div"
            >
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 10,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <h2
                  className="pageTitle"
                >
                  Document manager
                </h2>
              </Col>
            </Row>
            <Row
              tag="div"
            >
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 4,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <SearchButton
                  filter=""
                  handleFilter={[Function]}
                />
              </Col>
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 4,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <UploadButton
                  handleUpload={[Function]}
                />
              </Col>
            </Row>
            <Row
              tag="div"
            >
              <Col
                sm={
                  Object {
                    "offset": 1,
                    "size": 10,
                  }
                }
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <List
                  documents={null}
                  filter=""
                  handleDelete={[MockFunction]}
                />
              </Col>
            </Row>
          </Container>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Row
                tag="div"
              >
                <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 10,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <h2
                    className="pageTitle"
                  >
                    Document manager
                  </h2>
                </Col>
              </Row>,
              <Row
                tag="div"
              >
                <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 4,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <SearchButton
                    filter=""
                    handleFilter={[Function]}
                  />
                </Col>
                <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 4,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <UploadButton
                    handleUpload={[Function]}
                  />
                </Col>
              </Row>,
              <Row
                tag="div"
              >
                <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 10,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <List
                    documents={null}
                    filter=""
                    handleDelete={[MockFunction]}
                  />
                </Col>
              </Row>,
            ],
            "tag": "div",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 10,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <h2
                    className="pageTitle"
                  >
                    Document manager
                  </h2>
                </Col>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <h2
                    className="pageTitle"
                  >
                    Document manager
                  </h2>,
                  "sm": Object {
                    "offset": 1,
                    "size": 10,
                  },
                  "tag": "div",
                  "widths": Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Document manager",
                    "className": "pageTitle",
                  },
                  "ref": null,
                  "rendered": "Document manager",
                  "type": "h2",
                },
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <Col
                    sm={
                      Object {
                        "offset": 1,
                        "size": 4,
                      }
                    }
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <SearchButton
                      filter=""
                      handleFilter={[Function]}
                    />
                  </Col>,
                  <Col
                    sm={
                      Object {
                        "offset": 1,
                        "size": 4,
                      }
                    }
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <UploadButton
                      handleUpload={[Function]}
                    />
                  </Col>,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <SearchButton
                      filter=""
                      handleFilter={[Function]}
                    />,
                    "sm": Object {
                      "offset": 1,
                      "size": 4,
                    },
                    "tag": "div",
                    "widths": Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "filter": "",
                      "handleFilter": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <UploadButton
                      handleUpload={[Function]}
                    />,
                    "sm": Object {
                      "offset": 1,
                      "size": 4,
                    },
                    "tag": "div",
                    "widths": Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "handleUpload": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Col
                  sm={
                    Object {
                      "offset": 1,
                      "size": 10,
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <List
                    documents={null}
                    filter=""
                    handleDelete={[MockFunction]}
                  />
                </Col>,
                "tag": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <List
                    documents={null}
                    filter=""
                    handleDelete={[MockFunction]}
                  />,
                  "sm": Object {
                    "offset": 1,
                    "size": 10,
                  },
                  "tag": "div",
                  "widths": Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "documents": null,
                    "filter": "",
                    "handleDelete": [MockFunction],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
